libflux/flux-core/src/semantic/version.rs

```diff
@@ -0,0 +1,139 @@
+//! Utility to compare Flux types with their version.
+
+use std::cmp::max;
+
+use crate::semantic::types::{MonoType, PolyType};
+
+/// VersionDiff represents a difference between two types.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum VersionDiff {
+    /// Represents a backwards incomptible change to the types
+    Major = 2,
+    /// Represents a backwards comptible change to type
+    Minor = 1,
+    /// Represents no change to the types
+    Patch = 0,
+}
+
+/// Report the version difference between the old and new polytypes.
+pub fn diff(old: &PolyType, new: &PolyType) -> VersionDiff {
+    if old == new {
+        return VersionDiff::Patch;
+    }
+
+    // TODO handle constraints
+
+    diff_monotype(&old.expr, &new.expr)
+}
+
+fn diff_monotype(old: &MonoType, new: &MonoType) -> VersionDiff {
+    if old == new {
+        return VersionDiff::Patch;
+    }
+    match (old, new) {
+        (MonoType::Collection(old), MonoType::Collection(new)) => {
+            if old.collection != new.collection {
+                VersionDiff::Major
+            } else {
+                diff_monotype(&old.arg, &new.arg)
+            }
+        }
+        (MonoType::Record(old), MonoType::Record(new)) => {
+            for ofield in old.fields() {
+                if let Some(nfield) = new.fields().find(|f| f.k == ofield.k) {
+                    if diff_monotype(&ofield.v, &nfield.v) == VersionDiff::Major {
+                        return VersionDiff::Major;
+                    }
+                } else {
+                    return VersionDiff::Major;
+                }
+            }
+            // TODO in some cases adding a new record field is breaking and in
+            // others its not. How do we handle this ambiquity?
+            for nfield in new.fields() {
+                if old.fields().find(|f| f.k == nfield.k).is_none() {
+                    // Adding a new field to a record is a major change
+                    return VersionDiff::Major;
+                }
+            }
+            VersionDiff::Minor
+        }
+        (MonoType::Dict(ref old), MonoType::Dict(ref new)) => {
+            match (
+                diff_monotype(&old.key, &new.key),
+                diff_monotype(&old.val, &new.val),
+            ) {
+                (VersionDiff::Major, _) | (_, VersionDiff::Major) => VersionDiff::Major,
+                (VersionDiff::Patch, VersionDiff::Patch) => VersionDiff::Patch,
+                (_, _) => VersionDiff::Minor,
```

Comment:
Could extract this into a function to reuse as I think you need to "merge" two `VersionDiff`s else where.

Reply:
Actually, this is just `max`, which I see you have used already.